app/org/sagebionetworks/bridge/services/AuthenticationService.java

```diff
@@ -217,9 +213,30 @@ public UserSession signIn(Study study, CriteriaContext context, SignIn signIn) t
         
         return session;
     }
+    
+    public UserSession reauthenticate(Study study, CriteriaContext context, SignIn signIn) throws EntityNotFoundException {
+        checkNotNull(study);
+        checkNotNull(context);
+        checkNotNull(signIn);
+        
+        Validate.entityThrowingException(SignInValidator.REAUTHENTICATION_REQUEST, signIn);
+
+        Account account = accountDao.reauthenticate(study, signIn);
+
+        // Force recreation of the session, including the session token
+        cacheProvider.removeSessionByUserId(account.getId());
+        UserSession session = getSessionFromAccount(study, context, account);
+        cacheProvider.setUserSession(session);
+        
+        if (!session.doesConsent() && !session.isInRole(Roles.ADMINISTRATIVE_ROLES)) {
+            throw new ConsentRequiredException(session);
+        }
+        return session;
+    }
 
     public void signOut(final UserSession session) {
         if (session != null) {
+            accountDao.signOut(session.getStudyIdentifier(), session.getParticipant().getEmail());
             cacheProvider.removeSession(session);
         }
```

Comment:
Add a unit test?

Reply:
Added both normal and null session cases.